src/data_gradients/feature_extractors/common/image_duplicates.py

```diff
@@ -0,0 +1,264 @@
+
+        train_data: Iterable, the train_data used by AbstractManager (not used at the moment, but acts as a placeholder that
+            will later be used to derrive the diretory automatically according to its type - i.e ImageFolder.root etc).
+
+        val_data, Iterbale, the val_Data used by AbstractManager.
+        """
+        # TODO: ADD AUTOMATIC EXTRACTION FOR SG DATASETS
+        if self.train_image_dir is None:
+            self.train_image_dir = self._get_image_dir("train")
+        if val_data is not None and self.valid_image_dir is None:
+            self.valid_image_dir = self._get_image_dir("validation")
+
+    def _get_image_dir(self, split: str) -> str:
+        p = input(f"Image duplicates extraction: please enter the full path to the directory containing all {split} images >>> \n")
+        if not os.path.exists(p):
+            raise ValueError(f"Path to the directory containing all {split} images does not exist.")
+        return p
+
+    def update(self, sample: ImageSample):
+        """
+        Passed as we only find the duplicates in self.aggregate (per sample updated logic does not fit here).
+        """
+        pass
+
+    def _find_duplicates(self):
+        """
+        Finds duplicates in self.train_image_dir, self.valid_image_dir (when present) and their intersection.
+        Populates self.train_dups, self.valid_dups, self.intersection_dups and the corresponding appearences count attributes
+        self.train_dups_appearences, self.validation_dups_appearences, self.intersection_train_appearnces, self.intersection_val_appearnces.
+        """
+        dhasher = DHash()
+
+        # encodings = { IMAGE_X_FNAME: IMAGE_X_ENCODING,...}
+        train_encodings = dhasher.encode_images(self.train_image_dir)
+        valid_encodings = dhasher.encode_images(self.valid_image_dir)
+
+        # ADD PATH PREFIXES, SO WE CAN DIFFER TRAIN/VALIDATION AFTER CALLING find_duplicates
+        # AND AVOID COLLISIONS IF SAME FNAMES ARE USED BETWEEN THE DATASETS
+        # encodings = { IMAGE_X_PATH: IMAGE_X_ENCODING,...}
+        train_encodings = {str(os.path.join(self.train_image_dir, k)): v for k, v in train_encodings.items()}
+        valid_encodings = {str(os.path.join(self.valid_image_dir, k)): v for k, v in valid_encodings.items()}
+
+        # PASS ALL ENCODINGS
+        all_encodings = {**train_encodings, **valid_encodings}
+
+        # dups = { IMAGE_X_PATH: [DUPLICATE_OF_IMAGE_X_PATH, DUPLICATE_2_OF_IMAGE_X_PATH...]...}
+        dups = dhasher.find_duplicates(encoding_map=all_encodings, max_distance_threshold=0)
+
+        # FILTER PATHS WHERE NO DUPLICATES WERE FOUND
+        dups = {k: v for k, v in dups.items() if len(v) > 0}
+
+        train_dups = []
+        valid_dups = []
+        intersection_dups = []
+        dup_clique_heads = list(dups.keys())
+
+        # ITERATE THROUGH THE CLIQUE 'HEADS' (I.E SOME MEMBER REPRESENTING THE DUPLICATE CLIQUE)
+        for i in range(len(dup_clique_heads)):
```

Comment:
there is actually an algorithm for that. it is called "union-find"
https://en.wikipedia.org/wiki/Disjoint-set_data_structure. 
But nevermind

Reply:
Forgive me, it has been a long time since my data structures course (: